scarlet2/nn.py

```diff
@@ -0,0 +1,156 @@
+# -------------------------------------------------- #
+# This class is used to calculate the gradient       #
+# of the log-probability via calling the nn prior    #
+# ScoreNet model. A custom vjp is created to return  #
+# the nn prior when calling jax.grad()               #
+# -------------------------------------------------- #
+
+try:
+    import numpyro.distributions as dist
+    import numpyro.distributions.constraints as constraints
+    from galaxygrad import HSC_ScoreNet32, HSC_ScoreNet64, ZTF_ScoreNet32, ZTF_ScoreNet64,HSC_LogNet32, HSC_LogNet64# (https://pypi.org/project/galaxygrad/0.0.19)
+
+except ImportError:
+    raise ImportError("scarlet2.nn requires numpyro and galaxygrad=0.0.19")
+
+import jax.numpy as jnp
+import jax.scipy as jsp
+from jax import custom_vjp
+from jax import vjp
+
+
+def pad_fwd(x, trained_model):
+    """Zero-pads the input image to the nearest 32 or 64 pixels"""
+    data_size = x.shape[1]
+    pad = True
+    
+    # select the HSC trained model
+    if trained_model in ('hsc', 'None'):
+        if data_size <= 32:
+            pad_gap = 32 - data_size
+            ScoreNet = HSC_ScoreNet32
+        else:
+            pad_gap = 64 - data_size
+            ScoreNet = HSC_ScoreNet64
+
+    
+    # select the ZTF trained model
+    elif trained_model == 'ztf':
+        if data_size <= 32:
+            pad_gap = 32 - data_size
+            ScoreNet = ZTF_ScoreNet32
+        else:
+            pad_gap = 64 - data_size
+            ScoreNet = ZTF_ScoreNet64
+    # select the custom trained model
+    else:
+        ScoreNet = trained_model
+    # dont pad if we dont need to
+    if pad_gap == 0:
+        pad = False
+    # calculate how much to pad    
+    if pad_gap % 2 == 0:
+        pad_lo = pad_hi = int(pad_gap / 2)
+    else:
+        pad_lo = int(pad_gap // 2)
+        pad_hi = int(pad_lo + 1)
+    # perform the zero-padding
+    if pad:
+        if jnp.ndim(x) == 3:
+            x[0] = jnp.pad(x[0], ((pad_lo, pad_hi), (pad_lo, pad_hi)), 'constant', constant_values=0) 
+        else:
+            x = jnp.pad(x, ((pad_lo, pad_hi), (pad_lo, pad_hi)), 'constant', constant_values=0)
+    return x, ScoreNet, pad_lo, pad_hi , pad
+    
+# reverse pad back to original size
+def pad_back(x, pad_lo, pad_hi):
+    """Removes the zero-padding from the input image"""
+    if jnp.ndim(x) > 2: 
+        x[0] = x[pad_lo:-pad_hi, pad_lo:-pad_hi]
+    else:
+        x = x[pad_lo:-pad_hi, pad_lo:-pad_hi]
+    return x
+
+# calculate score function (jacobian of log-probability)
+def calc_grad(x, trained_model):
```

Comment:
let's rename `trained_model` with `model`, by which we mean the actual jax model that computes the score (instead of its name)

Reply:
renamed